### PR TITLE
Migrate issues comment WS to java

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/issue/IssueTesting.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      .setType(RuleType.CODE_SMELL)
+      .setComponent(file)
+      .setProject(project)
+      .setStatus(Issue.STATUS_OPEN)
+      .setResolution(null)
+      .setSeverity(Severity.MAJOR)
+      .setEffort(10L)
+      .setIssueCreationDate(DateUtils.parseDate("2014-09-04"))
+      .setIssueUpdateDate(DateUtils.parseDate("2014-12-04"))
+      .setCreatedAt(1400000000000L)
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueTesting.java]: MINOR Add underscores to this numeric value for readability [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2148)

File: sonar-db/src/main/java/org/sonar/db/issue/IssueTesting.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      .setUpdatedAt(1400000000000L);
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueTesting.java]: MINOR Add underscores to this numeric value for readability [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2148)

File: sonar-db/src/main/java/org/sonar/db/issue/IssueTesting.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import org.sonar.api.issue.Issue;
+import org.sonar.api.rule.Severity;
+import org.sonar.api.rules.RuleType;
+import org.sonar.api.utils.DateUtils;
+import org.sonar.core.util.Uuids;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.rule.RuleDto;
+
+public class IssueTesting {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueTesting.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1118)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 1 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static org.sonar.core.util.Uuids.UUID_EXAMPLE_01;
+import static org.sonarqube.ws.client.issue.IssuesWsParameters.ACTION_DELETE_COMMENT;
+import static org.sonarqube.ws.client.issue.IssuesWsParameters.PARAM_COMMENT;
+
+public class DeleteCommentAction implements IssuesWsAction {
+
+  private final System2 system2;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java]: MAJOR Remove this unused "system2" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/EditCommentAction.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    this.dbClient = dbClient;
+    this.issueFinder = issueFinder;
+    this.responseWriter = responseWriter;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_EDIT_COMMENT)
+      .setDescription("Edit a comment.<br/>" +
+        "Requires authentication and Browse permission on project of the specified issue<br/>" +
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/EditCommentAction.java]: suggestion: same modification on the permission description than add comment

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/EditCommentAction.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    action.createParam(PARAM_COMMENT)
+      .setDescription("Comment key")
+      .setDeprecatedKey("key")
+      .setSince("6.3")
+      .setRequired(true)
+      .setExampleValue(UUID_EXAMPLE_01);
+    action.createParam(PARAM_TEXT)
+      .setDescription("Comment text")
+      .setRequired(true)
+      .setExampleValue("blabla...");
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/EditCommentAction.java]: suggestion: more meaningful comment

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/EditCommentAction.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  @Override
+  public void handle(Request request, Response response) {
+    userSession.checkLoggedIn();
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      IssueDto issueDto = Stream.of(request)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/EditCommentAction.java]: 🎆

File: server/sonar-server/src/test/java/org/sonar/server/issue/ws/EditCommentActionTest.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+public class EditCommentActionTest {
+
+  private static final long NOW = 10_000_000_000L;
+
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  @Rule
+  public DbTester dbTester = DbTester.create(System2.INSTANCE);
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/issue/ws/EditCommentActionTest.java]: no need for the System2.INSTANCE

File: server/sonar-server/src/test/java/org/sonar/server/issue/ws/EditCommentActionTest.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    assertThat(action.key()).isEqualTo("edit_comment");
+    assertThat(action.isPost()).isTrue();
+    assertThat(action.isInternal()).isFalse();
+    assertThat(action.params()).hasSize(2);
+    assertThat(action.responseExample()).isNull();
+  }
+
+  private TestResponse call(@Nullable String commentKey, @Nullable String commentText) {
+    TestRequest request = tester.newRequest();
+    if (commentKey != null) {
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/issue/ws/EditCommentActionTest.java]: suggestion: setNullable

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    this.dbClient = dbClient;
+    this.issueFinder = issueFinder;
+    this.responseWriter = responseWriter;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_DELETE_COMMENT)
+      .setDescription("Delete a comment.<br/>" +
+        "Requires authentication and Browse permission on project of the specified issue<br/>" +
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java]: you just need 'Browse' to delete a comment?

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+      .setSince("6.3")
+      .setRequired(true)
+      .setExampleValue(UUID_EXAMPLE_01);
+  }
+
+  @Override
+  public void handle(Request request, Response response) {
+    userSession.checkLoggedIn();
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      IssueDto issueDto = Stream.of(request)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java]: 😀

File: server/sonar-server/src/test/java/org/sonar/server/issue/ws/DeleteCommentActionTest.java [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+import static org.sonar.api.web.UserRole.CODEVIEWER;
+import static org.sonar.api.web.UserRole.USER;
+
+public class DeleteCommentActionTest {
+
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  @Rule
+  public DbTester dbTester = DbTester.create(System2.INSTANCE);
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/issue/ws/DeleteCommentActionTest.java]: remark: without System2.INSTANCE

File: sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java [hunk 13]
```diff
@@ -93,17 +93,11 @@ public void insert(DbSession session, IssueChangeDto change) {
...
-      return count == 1;
-
-    } finally {
-      MyBatis.closeQuietly(session);
-    }
+  public boolean delete(DbSession session, String key) {
+    IssueChangeMapper mapper = mapper(session);
+    int count = mapper.delete(key);
+    session.commit();
+    return count == 1;
```
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java]: should we use this pattern?

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    this.issueFieldsSetter = issueFieldsSetter;
+    this.responseWriter = responseWriter;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(IssuesWsParameters.ACTION_ADD_COMMENT)
+      .setDescription("Add a comment.<br/>" +
+        "Requires authentication and Browse permission on project of the specified issue")
+      .setSince("3.6")
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: As discussed, almost. I will add a comment saying that the result is no more the same, it's now returning the issue.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    String issueKey = request.mandatoryParam(PARAM_ISSUE);
+    String commentText = request.mandatoryParam(PARAM_TEXT);
+    checkArgument(!isNullOrEmpty(commentText), "Cannot add empty comment to an issue");
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      IssueDto issueDto = issueFinder.getByKey(dbSession, issueKey);
+      IssueChangeContext context = IssueChangeContext.createUser(new Date(system2.now()), userSession.getLogin());
+      DefaultIssue defaultIssue = issueDto.toDefaultIssue();
+      issueFieldsSetter.addComment(defaultIssue, commentText, context);
+      issueUpdater.saveIssue(dbSession, defaultIssue, context, commentText);
+      responseWriter.write(issueKey, request, response);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: Why ? It's not JOSN, it's protobuff

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java (hunk 10 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/DeleteCommentAction.java]: Authentication + Browse = As you can only delete your own comment I think it's ok

File: sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java (hunk 13 above)
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/issue/IssueChangeDao.java]: It's used in bulk I think, but I agree I'd like to remove it.
Unfortunately not enough time to investigate for the moment.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  public void handle(Request request, Response response) {
+    userSession.checkLoggedIn();
+    String issueKey = request.mandatoryParam(PARAM_ISSUE);
+    String commentText = request.mandatoryParam(PARAM_TEXT);
+    checkArgument(!isNullOrEmpty(commentText), "Cannot add empty comment to an issue");
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      IssueDto issueDto = issueFinder.getByKey(dbSession, issueKey);
+      IssueChangeContext context = IssueChangeContext.createUser(new Date(system2.now()), userSession.getLogin());
+      DefaultIssue defaultIssue = issueDto.toDefaultIssue();
+      issueFieldsSetter.addComment(defaultIssue, commentText, context);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: I disagree, as other methods really modify the here, here a new comment is added

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  @Override
+  public void handle(Request request, Response response) {
+    userSession.checkLoggedIn();
+    String issueKey = request.mandatoryParam(PARAM_ISSUE);
+    String commentText = request.mandatoryParam(PARAM_TEXT);
+    checkArgument(!isNullOrEmpty(commentText), "Cannot add empty comment to an issue");
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      IssueDto issueDto = issueFinder.getByKey(dbSession, issueKey);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: It's a personal opinion, for me it's the same

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: I understand but it's not just a personal opinion. What we call a DTO is not a DTO. See http://stackoverflow.com/a/1612671/5052331 for an explanation

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: Then please do a proposition to rename the name of all DTO classes :)

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/AddCommentAction.java]: I already had some discussions about it and tried to push for it

**Pull request merged**